src/dpdb.js

```diff
@@ -0,0 +1,200 @@
+ * of the DPDB and updates the device info if a better match is found.
+ * Calls the onDeviceParamsUpdated callback when there is an update to the
+ * device information.
+ */
+function Dpdb(fetchOnline, onDeviceParamsUpdated) {
+  // Start with the offline DPDB cache while we are loading the real one.
+  this.dpdb = DPDB_CACHE;
+
+  // Calculate device params based on the offline version of the DPDB.
+  this.recalculateDeviceParams_();
+
+  // XHR to fetch online DPDB file, if requested.
+  if (fetchOnline) {
+    // Set the callback.
+    this.onDeviceParamsUpdated = onDeviceParamsUpdated;
+
+    console.log("Fetching DPDB...");
+    var xhr = new XMLHttpRequest();
+    var obj = this;
+    xhr.open("GET", ONLINE_DPDB_URL, true);
+    xhr.onreadystatechange = function() {
+      if (xhr.readyState != 4) return;
+
+      obj.loading = false;
+      if (xhr.status >= 200 && xhr.status <= 299) {
+        // Success.
+        console.log("Successfully loaded online DPDB.");
+        obj.dpdb = JSON.parse(xhr.response);
+        obj.recalculateDeviceParams_();
+      } else {
+        // Error loading the DPDB.
+        console.error("Error loading online DPDB!");
+      }
+    };
+    xhr.send();
+  }
+}
+
+// Returns the current device parameters.
+Dpdb.prototype.getDeviceParams = function() {
+  return this.deviceParams;
+};
+
+// Recalculates this device's parameters based on the DPDB.
+Dpdb.prototype.recalculateDeviceParams_ = function() {
+  console.log("Recalculating device params.");
+  var newDeviceParams = this.calcDeviceParams_();
+  console.log("New device parameters:");
+  console.log(newDeviceParams);
+  if (newDeviceParams) {
+    this.deviceParams = newDeviceParams;
+    // Invoke callback, if it is set.
+    if (this.onDeviceParamsUpdated) {
+      this.onDeviceParamsUpdated(this.deviceParams);
+    }
+  } else {
+    console.error("Failed to recalculate device parameters.");
+  }
+};
+
+// Returns a DeviceParams object that represents the best guess as to this
+// device's parameters. Can return null if the device does not match any
+// known devices.
+Dpdb.prototype.calcDeviceParams_ = function() {
+  var db = this.dpdb; // shorthand
+  if (!db) {
+    console.error("DPDB not available.");
+    return null;
+  }
+  if (db.format != "1") {
+    console.error("DPDB has unexpected format version.");
+    return null;
+  }
+  if (!db.devices || !db.devices.length) {
+    console.error("DPDB does not have a devices section.");
+    return null;
+  }
+
+  // Get the actual user agent and screen dimensions in pixels.
+  var userAgent = navigator.userAgent || navigator.vendor || window.opera;
+  var width = Util.getScreenWidth();
+  var height = Util.getScreenHeight();
+  console.log("User agent: " + userAgent);
+  console.log("Pixel width: " + width);
+  console.log("Pixel height: " + height);
+
+  for (var i = 0; i < db.devices.length; i++) {
+    var device = db.devices[i];
+    if (!device.ru) {
```

Comment:
I'm guessing you want ".ru" instead of ".rule" because it saves data in the JSON transport, but it's not really readable.